### PR TITLE
Migrate layers to React

### LINKED ISSUE #2310: Move layer components to React
Follow-up on https://github.com/tmrowco/electricitymap-contrib/issues/1681#issuecomment-572014604.

Moving layers to React and having them managed in one place in the DOM tree would be simpler and might help us optimizing their render performance.

#### Layers

* [ ] ExchangeLayer
* [ ] SolarLayer
* [ ] WindLayer

### PR DESCRIPTION
**Pull request by fbarl**
Resolves #2310.

This PR migrates all 3 map layers to React (to various degrees), removing further parts of DOM management from `main.js`, bringing us one step closer in addressing https://github.com/tmrowco/electricitymap-contrib/issues/2311, leaving only https://github.com/tmrowco/electricitymap-contrib/issues/2309 as the final step to complete it.

As an intermediary step, I made `zoneMap` a global object that is consumed by all the layers for getting the projection maps and zoom levels. As we address https://github.com/tmrowco/electricitymap-contrib/issues/2309, we'll be able to get rid of the global object in favor of passing props if we embed layers as children of the map component.

##### Changes

* Fully migrated to the React component
* Moved `MapExchangeTooltip` into the layer, bringing #2288 to near completion
* Created `exchangeQuantizedIntensityScale` with explicit carbon steps that map into the corresponding arrow images

##### Didn't do

* The map zooming is not working now when hovering over exchange arrows as the `ExchangeLayer` is not mapped deep enough in the map DOM anymore so the scroll events don't bubble up. I don't know of an elegant way to fix this before addressing https://github.com/tmrowco/electricitymap-contrib/issues/2309, so I'd fix it there if that's ok.
* ~Decided not to hide or stop the arrows animation when _dragging_ the map as the performance seems good even without it (the arrows are still hidden when they are outside of the viewport)~
* Not sure if performance is still a problem (see https://github.com/tmrowco/electricitymap-contrib/issues/1590), I didn't experience any lags on my machine

##### Changes

* Fully migrated to React (keeping only some canvas image filter helpers outside)
* Different behavior when dragging on the map - instead of caching the canvas state, the layer smoothly fades out/in when the map dragging starts/ends
* Cut down on a few steps of canvas image transformation, making it refresh a bit faster
* Using `solarColor` scale directly when generating the layer canvas image

##### Didn't do

* Using WebGL here would probably make this _a lot_ faster, probably recalculation could be done with a high FPS so we wouldn't even need to _freeze_ it during map dragging + we'd get the blur effect out of the box instead of depending on that lengthy helper - https://github.com/tmrowco/electricitymap-contrib/issues/923

##### Changes

* Moved the layer to React, still keeping it as a wrapper around _Windy_ (which still makes it a bit slow and clunky)
* Different behavior when dragging on the map - instead of caching the canvas state, the layer smoothly fades out/in when the map dragging starts/ends

##### Didn't do

* Using WebGL here would probably make this _a lot_ faster, probably recalculation could be done with a high FPS so we wouldn't even need to _freeze_ it during map dragging - we should probably merge `Windy` into the wind layer component as we migrate it to WebGL - https://github.com/tmrowco/electricitymap-contrib/issues/944

### CODE REVIEW

**Comment by corradio** [comment]
Hi @fbarl,

Thanks for this! Regarding performance, have you tested on very old devices?
It might be fast on your laptop but when I tested with old iPhones / iPads I had some lags especially during dragging. I think it would make sense to keep hiding the arrows while dragging if it is not too much work.

File: web/src/helpers/image.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
+/* eslint-disable */
```
Comment (corradio) [inline web/src/helpers/image.js:1]: I would put this in the solar component folder so we keep the solar component self-contained?

Comment (fbarl) [reply web/src/helpers/image.js:1]: I can do it if you want, but beware that `WindLayer` also has an external dependency `windy.js` and e.g. `AreaGraph` imports various methods from `helpers/*.js` :slightly_smiling_face: 

My idea here was that we'd be able to get rid of this file completely once we use e.g. WebGL blur filters as a part https://github.com/tmrowco/electricitymap-contrib/issues/923 and in the meantime I wanted to keep a clean separation between our internal `SolarLayer` React component and the temporary "external" blur helper which seems to be an adaptation of https://github.com/flozz/StackBlur and was never properly integrated in our code.

If you expect we'll be depending on that helper for a while, perhaps I can actually remove it in favor of using https://github.com/flozz/StackBlur directly, that would also somewhat simplify the code in `solarlayer.js`?

**Comment by fbarl** [comment]
> Regarding performance, have you tested on very old devices?
It might be fast on your laptop but when I tested with old iPhones / iPads I had some lags especially during dragging. I think it would make sense to keep hiding the arrows while dragging if it is not too much work.

Good point! I didn't test on the old devices, I just didn't notice a _significant_ downgrade in performance compared to production on my machine, but you're right that's not a good way to test performance.

I added the code to hide arrows when dragging in https://github.com/tmrowco/electricitymap-contrib/pull/2395/commits/fb84c8e866e6f857b76d553859452c6bd9a715aa, the change was straightforward.

PTAL :slightly_smiling_face:

File: web/src/helpers/image.js (hunk 1 above)
Comment (corradio) [reply web/src/helpers/image.js:1]: hmm let's keep it here for now. We're getting a bit pressed for time so I don't want to focus too much on these aspects which can be changed later as well :)

**Pull request merged**